src/utils/dplasma_lapack_adtt.c

```diff
@@ -76,116 +78,137 @@ static int dplasma_set_dtt_to_info( const dplasma_data_collection_t *dc, parsec_
-        if( NULL != dtt_entry ) {
-            assert(dtt_entry->adt.opaque_dtt == adt.opaque_dtt);
-            return 0;
+        if(parsec_atomic_cas_ptr(&dplasma_datatypes_lapack_helper, NULL, new_ht)) {
+            parsec_context_at_fini(dplasma_datatypes_info_fini, NULL);
+        } else {
+             parsec_hash_table_fini(new_ht);
+             PARSEC_OBJ_RELEASE(new_ht);
         }
     }
 
+    parsec_hash_table_lock_bucket(dplasma_datatypes_lapack_helper, k);  /* protect access to the hash table */
+    /* Find the entry */
+    dtt_entry = (dplasma_datatype_lapack_helper_t *)parsec_hash_table_nolock_find(dplasma_datatypes_lapack_helper, k);
+    if( NULL != dtt_entry ) {
+        assert(dtt_entry->adt.opaque_dtt == adt.opaque_dtt);
+        parsec_hash_table_unlock_bucket(dplasma_datatypes_lapack_helper, k);
+        return 0;
+    }
+
     dtt_entry = (dplasma_datatype_lapack_helper_t *)malloc(sizeof(dplasma_datatype_lapack_helper_t));
     dtt_entry->ht_item.key = (uint64_t)strdup(static_desc);
-    dtt_entry->info = info;
+    dtt_entry->info = *info;
     dtt_entry->adt = adt;
     parsec_hash_table_nolock_insert(dplasma_datatypes_lapack_helper, &dtt_entry->ht_item);
+    parsec_hash_table_unlock_bucket(dplasma_datatypes_lapack_helper, k);
     PARSEC_DEBUG_VERBOSE(27, parsec_debug_output,
         " insert dtt <- info lda %d rows %d cols %d loc %d shape %d layout %d dtt %p arena %p (%30s)",
-        info.lda, info.rows, info.cols, info.loc, info.shape, info.layout, adt.opaque_dtt, adt.arena, static_desc);
+        info->lda, info->rows, info->cols, info->loc, info->shape, info->layout, adt->opaque_dtt, adt->arena, static_desc);
     return 0;
 }
 
-int dplasma_set_datatype_info( const dplasma_data_collection_t *dc, parsec_arena_datatype_t adt,
-                               lapack_info_t info)
+int dplasma_set_datatype_info( const dplasma_data_collection_t *dc,
+                               parsec_arena_datatype_t adt,
+                               const lapack_info_t* info)
 {
     dplasma_set_dtt_to_info(dc, adt, info);
     dplasma_set_info_to_dtt(dc, adt, info);
     return 0;
 }
 
 int dplasma_get_info_from_datatype( const dplasma_data_collection_t *dc, parsec_datatype_t dtt,
-                                    lapack_info_t *info,
-                                    parsec_arena_datatype_t **adt)
+                                    const lapack_info_t **info,
+                                    const parsec_arena_datatype_t **adt)
 {
     dplasma_datatype_lapack_helper_t * dtt_entry = NULL;
     char static_desc[LAPACK_ADT_KEY_STR_SZ];
     snprintf( static_desc, LAPACK_ADT_KEY_STR_SZ, "%p|%p|-1|-1|-1", dc, dtt);
     parsec_key_t k = (uint64_t)static_desc;
     assert(dplasma_datatypes_lapack_helper!= NULL);
+    parsec_hash_table_lock_bucket(dplasma_datatypes_lapack_helper, k);  /* protect access to the hash table */
```

Comment:
Why not use `parsec_hash_table_find` here? Also, is there no chance that the `dtt_entry` can be removed from the hash table while we have the `info` pointer?

Reply:
I guess I could have used the `parsec_hash_table_find` and have the locks provided automatically. The info is related to a datatype/arena that is in use by this taskpool, so if should be refcounted at least until this taskpool completes.

Reply:
In the line below, we assume dtt_entry remains valid until we do the assignments `*info=dtt_entry->info` and `*adt =&dtt_endy->add`; 

Is it possible for a dtt_entry to be remove after a locked_find, but before the assignments? 

The second assignment also assumes that the dtt_entry will remain valid after the lock is released (as it points to the field in the dtt_entry. Is that correct?

Reply:
the bucket is locked, it cannot be removed until we release the lock down at line 175.